### PR TITLE
[OSDEV-3440] Add the address batch tool a private-instance admin can run themselves

### PR DESCRIPTION
**Pull request by daniellekain**
## Jira Ticket

[OSDEV-3440](https://opensupplyhub.atlassian.net/browse/OSDEV-3440)

---

## Summary of Changes

Submitting an address to OS Hub does not change what a production location displays — a submission is a contribution, and one contribution has to be **promoted** before its name and address become the location's primary values. Through the interface that is four operations per record, which does not scale to a cleanup of any size.

This graduates the design artifact from the OSDEV-3391 spike into a tool an RBA administrator can run **without us**, which is the requirement that drove the shape of it.

**Why a standalone API client and not a management command.** A management command would be more testable and would need no match discovery, and I initially proposed it. It is the wrong answer here: running one requires `ecs run-task` against OS Hub infrastructure, which the partner cannot do. The whole point is that they can run this themselves, so it has to go through the public API with their own superuser token.

**What changed from the artifact:**

- Configuration moved out of module globals read at import time into an injectable `Config`. This is what makes the logic testable at all, and it is most of the diff's value.
- **40 unit tests** with mocked transport.
- The created `match_id` is now read from the approval response when present, falling back to discovery. See the sequencing note below.
- The host guard is split into a pure function, so the precedence rule — a forbidden marker beats an allowed one — is actually covered by a test rather than assumed.
- An operator README written for someone outside our team.

**It also unbreaks the shellcheck job.** `run-bash-script-linter` runs `shellcheck scripts/*`, which fails outright the moment a *directory* exists under `scripts/` — `openBinaryFile: inappropriate type (is a directory)`. It is a required check, so as things stood no subdirectory could be added under `scripts/` at all, by anyone. Both the workflow step and the local `scripts/run_bash_script_linter` runner now lint files only, at depth one. Coverage is unchanged: every script directly in `scripts/` is still checked, including the extensionless ones a `*.sh` glob would have quietly dropped.

**The input path is validated before it is opened,** so a typo, a missing file or a directory produces a message naming the file rather than a traceback. The operator types that path, and a traceback is a poor first experience of a tool built to be run without us.

**Safety properties kept:** dry-run, per-location serial processing, a resumable JSONL journal in which only a *verified* record counts as done (so a run that died between approve and promote retries rather than skipping), a guard refusing to run against public OS Hub, and a per-record verification report a human reads before the batch is declared complete.

**Forward-compatible with OSDEV-3424.** The approval response carries only `os_id` today, which is why the match has to be discovered by taking our newest match for the location — correct only while nothing else writes to it concurrently. That assumption was defensible when *we* ran the batch; it is weaker when a partner runs it unsupervised. The tool now prefers a `match_id` in the approval response whenever one is there, so it is correct whichever order this and 3424 ship in, and stops inferring the moment 3424 deploys — with no change needed here.

---

## Type of Change

- [x] New feature (non-breaking change that adds functionality)
- [x] Documentation / tooling only

No application code is touched and nothing here executes as part of the application. Two files outside the tool are changed: the `run-bash-script-linter` workflow step and the local `scripts/run_bash_script_linter` runner, both for the shellcheck fix above. Those do execute — in CI — so they are the part of this diff worth a careful look.

---

## Testing

40 unit tests, `python -m unittest test_rba_address_batch` from the script directory — all passing. `flake8` clean on both files. The shellcheck change was verified by running the new invocation against this branch in a shellcheck container, with the new subdirectory present: clean, exit 0. `run-bash-script-linter` now passes on this PR.

Covered: the host guard including a public-instance host that also carries an allowed marker; submission payload with and without coordinates; the 422 duplicate-request window raising a distinct error; `match_id` taken from the approval response with no discovery call made; discovery falling back correctly, taking the newest of *our own* matches and ignoring another contributor's newer one; discovery raising when none of our matches are present; journal resume treating only `verified` as done; row grouping preserving file order.

Transport is mocked throughout — no test talks to a real instance.

**Not exercised against the RBA instance.** That is the point of the ticket's pre-flight items and cannot be done from CI.

---

## Known TODO Items

- **These tests do not run in CI.** The tool is a standalone client living outside the Django app, so the Django test runner does not collect it, and `scripts/` is an unmapped path in the code-quality workflow's path filter. They run locally with one command and are documented in the README. Wiring a job for them is a follow-up worth doing if more tooling lands here.
- **Three pre-flight items remain unverified against the instance**, per the ticket: that the superuser token authenticates, that an `ApiLimit` row exists for the contributor, and the base URL. The first real run is the verification. Note the `ApiLimit` row is the one prerequisite the partner cannot set up themselves.
- **OSDEV-3438 must be deployed before a batch is run** — not before this merges. Nothing here executes on merge, so the dependency is operational rather than a code one. On any instance receiving the one-way sync, promotions this tool applies are reverted on the next public change to those records unless the re-assert command is in place. That prerequisite was recorded only on this pull request, which is not where the person running the tool looks, so it is now the first thing in the README's "Before you start". The one caveat for release planning: this must not ship in an *earlier* release than 3438.
- **The concurrency caveat is documented, not enforced.** The README asks the operator not to edit the same locations or run two batches at once. OSDEV-3424 removes the need for the request.
- `requests` is the only dependency and is installed into a venv by the operator; it is deliberately not added to any application requirements file, since nothing in the app imports this.

---

### Implementation

- [x] My code follows the project's style guidelines and naming conventions.
- [x] I have removed debug logs, commented-out code, and unused imports.
- [x] Any AI-assisted code (e.g., Claude) has been reviewed, understood, and adapted to fit our codebase.
- [x] I have considered backward compatibility and migrations where applicable.
- [x] The diff contains no secrets, internal document/spreadsheet/folder IDs, internal board or channel identifiers, internal policy thresholds, or staff contact details — internal values are externalized to runtime configuration (see "Public repository hygiene" in AGENTS.md). Host and credentials are environment variables; the README uses placeholders and names no real host.

### Testing & Validation

- [x] I have manually tested the change in a local/dev environment.
- [x] I have added or updated unit tests for the new/changed behavior.
- [x] All existing tests pass locally.
- [ ] For backend changes: I have validated API contracts, error handling, and edge cases. — error handling and edge cases are covered by tests; the **API contracts are validated against the code, not against a running instance**, which is what the pre-flight items above cover.

### Documentation & Communication

- [x] I have updated relevant documentation (release notes, plus an operator README).
- [x] I have updated the Jira ticket status and added any relevant notes.

### Final Review

- [x] I have performed a self-review on my PR and I am presenting my best work for my peers to review.

[OSDEV-3440]: https://opensupplyhub.atlassian.net/browse/OSDEV-3440?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by valencia-max** [comment]
I found these issues using Claude:

Two correctness concerns in rba_address_batch.py, both around retry/resume:

1. Retry after any failure re-runs the whole submit→approve→promote sequence, which can duplicate mutations

execute() wraps the full process_row() call in one try/except, and completed_row_keys() only treats a 'verified' journal entry as done (rba_address_batch.py:294, 582-617). So:

If only the final read-back (fetch_primary after promote, line 562) fails — a pure read with no side effects — the row is still marked FAILED. Re-running calls submit_address again from scratch, creating a second moderation event/approval/promote for a location that's already correct.
If the run dies between submit and promote, the original submission is never approved, and the journal's recorded moderation_id is never reused to resume it. A later re-run (as the README instructs, after waiting out the 15-min duplicate window) creates a brand-new submission while the first sits orphaned, unapproved, forever — with nothing counting or surfacing these leftovers.

Suggest: use the journaled step/IDs to resume mid-record instead of always restarting from submit, and/or don't mark a row FAILED if everything except the trailing verification read succeeded.

2. Dry run's safety checks bypass the rate-limit backoff, so they can misreport under their own documented load

check_superuser_access() (line 479) and the per-location existence loop in dry_run() (lines 529-537) call session.get() directly instead of request_with_backoff(), unlike every other call in the tool. Given the docstring's own estimate ("roughly a thousand records... several thousand calls"), a large dry run issuing one GET per location risks hitting the same rate limit, which would surface as PROBLEM {os_id}: HTTP 429 — read as "location missing" — instead of being retried. That undermines the exact pre-flight check the dry run exists to provide.

Suggest: route both through request_with_backoff() for consistency.



Let me know what you think!!

**Comment by daniellekain** [comment]
This feedback was really helpful, thank you! I think I addressed both of the points in my most recent commit, lmk what you think. For 1. I moved the verification read outside the mutation path; for 2. I added the request_with_backoff and made the checks more stringent.

File: scripts/rba_address_batch/rba_address_batch.py [hunk 1]
```diff
@@ -0,0 +1,954 @@
...
+    first_label = hostname.split('.')[0]
+    if first_label in PRIVATE_INSTANCE_LABELS:
+        return None
+
+    return (
+        'refusing to run against {}: this tool is for a single-tenant '
+        'private instance ({}), or local development'.format(
+            hostname, ', '.join(sorted(PRIVATE_INSTANCE_LABELS))
+        )
+    )
```
[the remark this answers is not in the document]
Comment (daniellekain) [reply scripts/rba_address_batch/rba_address_batch.py]: Good catch, fixed in 81a7d184. The guard read the hostname and ignored the scheme, so http://rba.opensupplyhub.org passed it and would have put the superuser token on the wire in the clear.

One deviation from the suggested diff: I put the scheme check after the public-instance test rather than before it, so http://opensupplyhub.org still reports as "refusing to run against the public instance" rather than as a scheme problem — the most dangerous target should keep the message that says why. Local development stays exempt, since there is no TLS there and nothing in between to intercept. Three tests added for the three cases.

File: scripts/rba_address_batch/rba_address_batch.py [hunk 2]
```diff
@@ -0,0 +1,954 @@
...
+    re-read of a value this tool has since changed.
+    """
+    facts = facts or {}
+    reached = facts.get('reached', set())
+
+    before = facts.get('previous_primary_address')
+    if 'started' not in reached:
+        # Cache-busted, so this is the value as it stands rather than the
+        # one the dry run left in the response cache.
+        _, before = fetch_primary(session, config, os_id, cache_buster=key)
```
[the remark this answers is not in the document]
Comment (daniellekain) [reply scripts/rba_address_batch/rba_address_batch.py:724]: Accurate, and it is a limitation I cannot close from this side.

Making the endpoint return the existing moderation_id for a retried submission needs an idempotency key on PATCH /api/v1/production-locations/{os_id}/. This tool is a standalone API client that a partner runs with their own token, so it cannot make the endpoint do that.

What it does instead: the journal records a "started" entry before the PATCH is sent, so a row that dies in that window is reported as INCOMPLETE naming the step it reached rather than vanishing. On the re-run the resubmission raises DuplicateSubmission with a message saying the earlier event exists but is unreachable from here, and that it has to be approved and promoted from the moderation queue by hand, or the window waited out. The README says the same under "If a run is interrupted".

So the failure is loud and a human can recover it, but the id genuinely cannot be recovered without server-side idempotency. Worth a follow-up ticket if we want the API to support it.

File: scripts/rba_address_batch/rba_address_batch.py [hunk 3]
```diff
@@ -0,0 +1,954 @@
...
+            'row_key': key,
+            'step': 'started',
+            'os_id': os_id,
+            'submitted_address': row['address'],
+            'previous_primary_address': before,
+        })
+
+    if 'submitted' in reached:
+        moderation_id = required_fact(
+            facts, 'moderation_id', key, 'submitted'
```
[the remark this answers is not in the document]
Comment (daniellekain) [reply scripts/rba_address_batch/rba_address_batch.py:738]: This is the case I deliberately chose to stop on, and I do not think the suggested reconciliation is available.

"Reconcile the authoritative event state" would mean reading the event back after the 410. The only endpoint for that is GET /api/v1/moderation-events/{id}/, which is served from OpenSearch (see retrieve in views/v1/moderation_events.py) and lags by the indexing cycle, so right after an approval it can still report PENDING. It is not authoritative at exactly the moment we would need it to be.

The fallback, "recover its created match", is discover_match_id, which takes the newest match for our contributor. If the event was approved that is the right match; if a moderator rejected it, no new match was created and we would promote an unrelated older contribution over a real location. Nothing in the /split/ response links a match back to the moderation event that created it, so ownership cannot be proven — which is the condition your own note says should keep the ambiguity failure.

Hence the loud stop: it names the event and tells the operator to finish or drop that record from the moderation queue. It is a narrow window (approve request sent, response never received) and a human-resolvable outcome, which I would take over a silent wrong promote. OSDEV-3424 does not change this either — it adds match_id to the approval response, which helps the normal path but not the case where we never received that response.

File: scripts/rba_address_batch/README.md [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+> **Do not run this against an instance that receives the one-way sync
+> from Open Supply Hub until promotion re-assertion is deployed there.**
+> On such an instance the sync overwrites every synced field of a shared
+> location, including the record of which contribution is primary. A
+> promotion this tool applies is therefore silently undone the next time
+> the public record changes — there is no schedule to it, so a batch can
+> look entirely correct for days and then decay. The `reassert_rba_promotions`
+> management command restores promotions after each sync and must be in
+> place first. If you are not certain it is deployed, ask the Open Supply
+> Hub team before running a batch rather than after.
```
[the remark this answers is not in the document]
Comment (daniellekain) [reply scripts/rba_address_batch/README.md:29]: Stale now. reassert_rba_promotions landed on main with #1280 (OSDEV-3438) and is on this branch since I merged main in — src/django/api/management/commands/reassert_rba_promotions.py. The prerequisite is verifiable from the repo as written.

**Review by valencia-max — APPROVED** [review]
